plugin/output/file/files.go

```diff
@@ -0,0 +1,82 @@
+package file
+
+import (
+	"sync"
+
+	"github.com/ozonru/file.d/logger"
+	"github.com/ozonru/file.d/pipeline"
+)
+
+// Plugins is an abstraction upon multiple file.Plugin, which helps reuse it.
+type Plugins struct {
+	// plugins contains plugs that exist from start of work.
+	plugins map[string]*Plugin
+	// dynamicPlugins contains plugs, that created dynamically during execution.
+	// they separated from plugins to avoid races and reduce locking complexity.
+	dynamicPlugins map[string]*Plugin
+	mu             sync.RWMutex
+}
+
+func NewFilePlugins(plugins map[string]*Plugin) *Plugins {
+	return &Plugins{plugins: plugins}
+}
+
+func (p *Plugins) Out(event *pipeline.Event, selector pipeline.PluginSelector) {
+	switch selector.CondType {
+	case pipeline.ByNameSelector:
+		p.mu.RLock()
+		defer p.mu.RUnlock()
```

Comment:
этот дефер сработает при выходи из функции, а не из выхода из switch case
сейчас оно работает, но может запутать в будущем
как вариант можно сделать
```
case ...:
  func() {
    p.mu.RLock()
    defer p.mu.RUnlock()
    
    p.plugins...
  }
```

Reply:
I assume, that RLock usage is pretty clear there and new conditions won't add complexity:
```go
switch selector.CondType {
	case pipeline.ByNameSelector:
		p.mu.RLock()
		defer p.mu.RUnlock()

		p.plugins[selector.CondValue].Out(event)
        case pipeline.BySecondSelector:
               pluginName := someFunc(selector.CondValue)

                p.mu.RLock()
		defer p.mu.RUnlock()

		p.plugins[pluginName].Out(event)
	case pipeline.UnknownSelector:
		logger.Fatalf("PluginSelector type didn't set for event: %v, selector: %v", event, selector)
	}
```
If it's complicated I can add func, like you suggested, but it seems too much.

Reply:
problem not in the switch, but the code afterwards
it's easy to do something like 
```go
func (p *Plugins) Out(event *pipeline.Event, selector pipeline.PluginSelector) {
	switch selector.CondType {
	case pipeline.ByNameSelector:
		p.mu.RLock()
		defer p.mu.RUnlock()

		p.plugin...
	}
	doSomeOtherStuff()
}
```

and it will not be so obvious, that `doSomeOtherStuff` still in lock